### PR TITLE
Add post-mortem of May 31th

### PR DESCRIPTION
**Pull request by saraycp**
About the `rack` gem version mismatch.

### CODE REVIEW

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+Our [reference server](https://build.opensuse.org) was offline for 27 minutes. No one was able to work with the API or user interface during that time. Other services depending on OBS (like [https://software.opensuse.org](https://software.opensuse.org)) were taken down by this as well.
+
+## Root Causes
+
+After the update of the `rack` ruby gem in our bundle, there was a mismatch with the version installed in the system. The error message we found in the log was:
+
+```
+You have already activated rack 2.2.3, but your Gemfile requires rack 2.2.3.1. Prepending `bundle exec` to your command may solve this. (Gem::LoadError)
+```
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:20]: I don't think this is the root cause, this is only part of what happened. 

The root cause is that we had different `rack` versions installed in the system and the bundle. Similar to the `strscan` incident.

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+## Trigger
+
+Merging and deploying the `rack` gem update.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:24]: Deployment was the trigger.

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+
+## Lessons Learned
+
+### What Went Well
+
+Collaboration among the team to resolve this.
+
+### What Went Wrong
+
+The server was down for too long.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:47]: That always is wrong in an incident, I don't think it's worth mentioning. Here are some other things I think went wrong.

- We did not resolve all the action items from the `strscan` incident yet (might have saved us)
- We had a handover for DemolitionSquad on Monday (A->B) and on Tuesday (B-A) that didn't go well
- We did not block deployment (technically) even though we were aware of the problem

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+
+### Where We Got Lucky
+
+No permanent damage or data loss.
+
+## Timeline (times in UTC)
+
+- 08:33 Start the deployment which isn't successful.
+- 08:37 Detect the server is down.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:56]: hm 4 minutes? Really?

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+- 08:40 Receive alerts from our monitoring system.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:57]: This also seems off. It took 7 minutes to receive an alert?

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+- 08:44 Inform on different channels.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:58]:
```suggestion
- 08:44 We informed people on different channels
```

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 7]
```diff
@@ -0,0 +1,65 @@
...
+- 08:51 Try to restart Apache.
+- 08:55 See the message on the Passenger log:
+        "You have already activated rack 2.2.3, but your Gemfile requires rack 2.2.3.1. Prepending `bundle exec` to your command may solve this. (Gem::LoadError)"
+- 08:58 Change the rack gem version on Gemfile.lock and restart Apache. This doesn't solve the problem.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:62]: This did solve the problem. We just did not wait long enough I think.

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: _posts/deployments/2022-05-31-post-mortem.md (hunk 4 above)
Comment (saraycp) [reply _posts/deployments/2022-05-31-post-mortem.md:56]: I added another item to the list to be more precise.

- Deployment started at 08:33
- we got the error in the output of `ansible-obs` at 08:35
- around 08:36/08:37 I took some screenshots of the output, checked the whole output trying to understand the error, and tried to access OBS which was down.
- we received the alerts at 08:40, the reason is that the alert checks if we are over/under the limit for 5 mins.

I attach two screenshots to demonstrate what I explain:

Screenshot from 2022-05-31 16-47-53

Screenshot 2022-05-31 at 16-46-43 Passenger - Grafana

Comment (hennevogel) [reply _posts/deployments/2022-05-31-post-mortem.md:56]: Okay then we have another TODO. We need a shorter alerting timeframe.

File: _posts/deployments/2022-05-31-post-mortem.md (hunk 7 above)
Comment (saraycp) [reply _posts/deployments/2022-05-31-post-mortem.md:62]: Rephrased

File: _posts/deployments/2022-05-31-post-mortem.md (hunk 3 above)
Comment (saraycp) [reply _posts/deployments/2022-05-31-post-mortem.md:47]: I omit the second point in the post-mortem, we already talked about that in the Retro meeting. Of course, this is important and we have to do our best in this regard, but I think that's a team matter, not interesting for the people reading the post-mortem.

File: _posts/deployments/2022-05-31-post-mortem.md (hunk 1 above)
Comment (saraycp) [reply _posts/deployments/2022-05-31-post-mortem.md:20]: Rephrased.

File: _posts/deployments/2022-05-31-post-mortem.md (hunk 4 above)
Comment (saraycp) [reply _posts/deployments/2022-05-31-post-mortem.md:56]: > Okay then we have another TODO. We need a shorter alerting timeframe.

I'm going to reduce the timeframe of the alerts involved in this case. I can change it from 5mins to 2mins maybe and we'll see if it is too much.

**Comment by hennevogel** [comment]
Pushed something on top

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+- On Tuesday we deployed without checking in with the previous operator
+
+### Where We Got Lucky
+
+No permanent damage or data loss.
+
+## Timeline (times in UTC)
+
+- 08:33 Start the deployment which isn't successful.
+- 08:35 Deployment failed.
```
Comment (hennevogel) [inline _posts/deployments/2022-05-31-post-mortem.md:65]: Another question I have. *What* exactly failed here?

**Review by hennevogel — CHANGES_REQUESTED** [review]

File: _posts/deployments/2022-05-31-post-mortem.md [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+
+## Resolution
+
+Updating the rack package (`zypper up ruby3.1-rubygem-rack`)
+
+## Action Items
+
+- Requiring a specific rack version in `obs-bundled-gems` again ([we removed this some time ago](https://github.com/openSUSE/open-build-service/commit/4d078f955b0c68d3753cd4fd293d0e7783fc5d41))
+- Fix our long standing OPS TODO with [blue/green deployments](https://en.wikipedia.org/wiki/Blue-green_deployment)
+-
```
[the remark this answers is not in the document]
Comment (saraycp) [reply _posts/deployments/2022-05-31-post-mortem.md]: I added a commit on top for this, it should be squashed.

**Comment by saraycp** [comment]
@hennevogel I added some commits on top, they should be squashed before merging.

File: _posts/deployments/2022-05-31-post-mortem.md (hunk 8 above)
Comment (hennevogel) [reply _posts/deployments/2022-05-31-post-mortem.md]: Okay so this and the next event in the timeline is the same thing.

**Comment by hennevogel** [comment]
@saraycp I squashed things, can you check again?

**Review by hennevogel — APPROVED** [review]

**Pull request merged**